### PR TITLE
Add semester database

### LINKED ISSUE #24: Semesters database
We need to create a table for storing semesters in our PostgreSQL database.

### PR DESCRIPTION
**Pull request by AdamBucior**
Closes #24

### CODE REVIEW

File: db/00-init.sql [hunk 1]
```diff
@@ -5,6 +5,14 @@ CREATE TABLE users (
     password_hash text
 );
 
+CREATE TABLE semesters (
+    id uuid PRIMARY KEY,
+    year integer NOT NULL,
+    type varchar(32) NOT NULL,
```
Comment (Trzcin) [inline db/00-init.sql:11]: Could we make this column an enum with `winter` and `summer` values or is there a good reason not to?

File: www/src/backend/semester.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
+import assert from 'node:assert';
+import { db } from '@backend/db';
+
+export type SemesterType = 'summer' | 'winter';
```
Comment (Trzcin) [inline www/src/backend/semester.ts:4]: Keep in mind that there is also the edge case of the 7th bachelor semester that is shorter than the regular winter one. The calendar on the university page makes a distinction for it so maybe we should as well? https://pg.edu.pl/studenci/studia/kalendarz-roku-akademickiego

File: www/src/backend/semester.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+export interface SemesterCreateData {
+    year: number;
+    type: SemesterType;
+    startDate: Date;
+    endDate: Date;
+}
+
+export interface SemesterEditData {
```
Comment (Trzcin) [inline www/src/backend/semester.ts:21]: We could make a type helper for this at this point since we have a similar type for users and will likely repeat this pattern for other db tables. The type helper would be similar to `Partial` in TS but would also add the `| undefined` part to all keys.

File: www/src/backend/semester.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        if (data.type !== undefined) {
+            this.type = data.type;
+        }
+        if (data.startDate !== undefined) {
+            this.startDate = data.startDate;
+        }
+
+        if (data.endDate !== undefined) {
+            this.endDate = data.endDate;
+        }
```
Comment (Trzcin) [inline www/src/backend/semester.ts:82]: I would personally shorten these to `this.field = data.field ?? this.field` but this is just a nitpick

File: www/tests/unit/semester.test.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    expect(semester1.startDate.getDate()).toBe(1);
+    expect(semester1).toHaveProperty('endDate', new Date('2025-01-30T00:00:00'));
+    expect(semester1.endDate.getFullYear()).toBe(2025);
+    expect(semester1.endDate.getMonth()).toBe(0);
+    expect(semester1.endDate.getDate()).toBe(30);
+
+    expect(await Semester.fetch(semester2.id)).toStrictEqual(semester2);
+    expect(await Semester.fetch(semester1.id)).toStrictEqual(semester1);
+    expect(sortById(await Semester.fetchAll())).toStrictEqual(sortById([semester1, semester2]));
+});
```
Comment (Trzcin) [inline www/tests/unit/semester.test.ts:86]: I think we should break this test up into smaller ones since here if any semester functionality fails it is not immediately clear from the console test output if creating, fetching or updating failed.
I would also test some scenarios with incorrect data that cannot be just expressed with Typescript, for example a test where creating a semester with a duplicate type/year pair should fail.

**Review by Trzcin — CHANGES_REQUESTED** [review]

File: www/src/backend/semester.ts (hunk 2 above)
Comment (AdamBucior) [reply www/src/backend/semester.ts:4]: The calendar on lab.kti.gda.pl does not make this distinction. Furthermore this would complicate things by allowing there to be 2 overlapping semesters.

File: www/tests/unit/semester.test.ts (hunk 5 above)
Comment (Trzcin) [reply www/tests/unit/semester.test.ts:86]: After talking about this I agree this should remain as a single test case since the expect calls depend on the operations in the entire test (like the creation of the semesters). So due to this running smaller chunks independently wouldn't work because they would have side effects.

File: www/src/backend/semester.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        const data = (await db.query<DbSemester>('SELECT * FROM semesters WHERE id = $1', [id])).rows[0];
+
+        return data ? new Semester(data) : null;
+    }
+
+    static async fetchAll(): Promise<Semester[]> {
+        const records = (await db.query<DbSemester>('SELECT * FROM semesters')).rows;
+
+        return records.map(record => new Semester(record));
+    }
```
Comment (Trzcin) [inline www/src/backend/semester.ts:53]: Do you think it would be a good idea for this function to return a sorted list of semesters (first item being the newest semester)? I'm thinking about it since in the UI we would typically want the semesters to be in order.

File: www/src/backend/semester.ts (hunk 3 above)
Comment (AdamBucior) [reply www/src/backend/semester.ts:21]: Nah

File: www/src/backend/semester.ts (hunk 4 above)
Comment (AdamBucior) [reply www/src/backend/semester.ts:82]: Nah

File: db/00-init.sql (hunk 1 above)
Comment (AdamBucior) [reply db/00-init.sql:11]: Nah

Comment (AdamBucior) [reply db/00-init.sql:11]: Ok, fine.

File: www/src/backend/semester.ts (hunk 6 above)
Comment (AdamBucior) [reply www/src/backend/semester.ts]: Done

**Review by Trzcin — APPROVED** [review]
We could add one more line to the test to check if creating a duplicate semester throws (you can copy the line from the users unit test) but other than that LGTM since we have the DB constraint.
